In [1]:
import pandas as pd
import numpy as np
import os

In [2]:
path = r'/Users/andrew/Documents/Data Analyst Course/Data Immersion/Achievement 4/Instacart Basket Analysis'

In [3]:
df_prods = pd.read_csv(os.path.join(path, 'Data', 'Original Data', 'products.csv'), index_col = False)

In [4]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [5]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [8]:
df_prods.shape

(49693, 5)

In [11]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [12]:
df_prods_clean.shape

(49677, 5)

In [13]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [14]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [15]:
df_prods_clean.shape

(49677, 5)

In [16]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [17]:
df_prods_clean_no_dups.shape

(49672, 5)

In [18]:
df_prods_clean_no_dups.to_csv(os.path.join(path, 'Data','Prepared Data', 'products_wrangled.csv'))

# 4.5 Consistency Checks Task

In [50]:
df_ords = pd.read_csv(os.path.join(path, 'Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [51]:
# Run the df.describe() function on your df_prods dataframe. Using your new knowledge about 
# how to interpret the output of this function,
# share in a markdown cell whether anything about the data looks off or should be investigated further.

In [53]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [54]:
# Max price seems very high for a supermarket shop, further analysis required.

In [55]:
# Check for mixed-type data in your df_ords dataframe.

In [56]:
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   Unnamed: 0              int64  
 1   order_id                int64  
 2   user_id                 int64  
 3   order_number            int64  
 4   orders_day_of_week      int64  
 5   order_time              int64  
 6   days_since_prior_order  float64
dtypes: float64(1), int64(6)
memory usage: 182.7 MB


In [29]:
# If you find mixed-type data, fix it. The column in question should contain observations of a single data type.

In [57]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

In [58]:
# Run a check for missing values in your df_ords dataframe.

In [59]:
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_time                     0
days_since_prior_order    206209
dtype: int64

In [60]:
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [61]:
df_ords_nan

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_time  days_since_prior_order  
0                 8                     NaN  
11               11                     NaN  
26               14                     NaN  
39               11                     NaN  
45               12                     NaN  
...             ...                     ...  
3420930          12                     NaN  
3420934          18                     NaN  
3421002          19                     NaN  
3421019          15                     NaN  
3421069          11                     NaN  

[206209 rows x 7 columns]

206209 missing values in days_since_prior_order column, could be due to high numbers of single time buyers rather than repeat customers.

This data is still important even if there are missing entries. One suggestion would be to flag if it's a first time buyer instead.

In [62]:
# Defining new df for cleaned data
df_ords_clean = df_ords

In [63]:
df_ords_clean['first_order'] = df_ords['days_since_prior_order'].isnull() == True

In [64]:
df_ords_clean.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           1   2398795        1             2                   3   
2           2    473747        1             3                   3   
3           3   2254736        1             4                   4   
4           4    431534        1             5                   4   

   order_time  days_since_prior_order  first_order  
0           8                     NaN         True  
1           7                    15.0        False  
2          12                    21.0        False  
3           7                    29.0        False  
4          15                    28.0        False

In [67]:
# Run a check for duplicate values in your df_ords data.

In [68]:
df_ords_dups = df_ords_clean[df_ords_clean.duplicated()]

In [69]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_time, days_since_prior_order, first_order]
Index: []

In [ ]:
# Identify any values above 100 in the prices column and assigns them a value of np.nan (from NumPy library)

In [ ]:
ords_prods_merge.loc[ords_prods_merge['prices'] >100, 'prices'] = np.nan

In [ ]:
ords_prods_merge['prices'].max()

In [70]:
# Address the duplicates using an appropriate method

In [71]:
df_ords_dups.shape

(0, 8)

In [ ]:
# Export your final, cleaned df_prods and df_ords data as “.csv” files in
# your “Prepared Data” folder and give them appropriate, succinct names.

In [72]:
df_ords_clean.to_csv(os.path.join(path, 'Data','Prepared Data', 'orders_cleaned.csv'))

In [73]:
df_prods_clean_no_dups.to_csv(os.path.join(path, 'Data','Prepared Data', 'products_wrangled.csv'))